planner/implementation/datasource.go

```diff
@@ -115,16 +114,32 @@ func NewTableScanImpl(ts *plannercore.PhysicalTableScan, cols []*expression.Colu
 
 // CalcCost calculates the cost of the table scan Implementation.
 func (impl *TableScanImpl) CalcCost(outCount float64, children ...memo.Implementation) float64 {
-	ts := impl.plan.(*plannercore.PhysicalTableScan)
+	selCost := 0.0
+	tsCost := 0.0
+	sessVars := impl.plan.SCtx().GetSessionVars()
+	var ts *plannercore.PhysicalTableScan
+	if sel, ok := impl.plan.(*plannercore.PhysicalSelection); ok {
```

Comment:
I prefer not to do this. Why not generate a new selection instead of the additional field `AllConds` in `DataSource`?

Reply:
The main problem is to accurately estimate the `RowCount` of the `PhysicalSelection`. It needs to call `tableStats.HistColl.Selectivity()` with `AllConds`.